R/type2.R

```diff
@@ -409,3 +422,41 @@ vec_is_subtype <- function(x, super, ..., x_arg = "", super_arg = "") {
 vec_implements_ptype2 <- function(x) {
   .Call(vctrs_implements_ptype2, x)
 }
+
+# Example usage of ptype2 restart. As there is no cast restart (yet),
+# it requires `x` and `y` to be convertible to the restarted ptype.
+with_ordered_restart <- function(expr) {
+  withCallingHandlers(
+    expr,
+    vctrs_error_ptype2 = function(cnd) {
+      x <- cnd[["x"]]
+      y <- cnd[["y"]]
+
+      restart <- FALSE
+
+      if (is.ordered(x)) {
+        restart <- TRUE
+        x <- as.character(x)
+      }
+      if (is.ordered(y)) {
+        restart <- TRUE
+        y <- as.character(y)
+      }
+
+      # Don't recurse and let ptype2 error keep its course
+      if (!restart) {
+        return(zap())
```

Comment:
Is `zap()` your own convention? I assume you could really return anything?

Reply:
Yes returning `zap()` is a reminder that the result is ignored and allows switching to `try_fetch()` seamlessly.